codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,348 @@
+                        public void run() {
+                            streamInfo.childChannel.streamTerminated = true;
+                            streamInfo.childChannel.unsafe().disconnect(
+                                streamInfo.childChannel.newPromise());
+                        }
+                    });
+                }
+            }
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayMessage goAway = new Http2GoAwayMessage(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            if (streamInfo != null) {
+                                streamInfo.childChannel.pipeline()
+                                    .fireUserEventTriggered(goAway.retain());
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Throwable t) {
+                ctx.invoker().invokeExceptionCaught(ctx, t);
+            }
+            ctx.fireUserEventTriggered(goAway);
+        }
+    }
+
+    class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        public InternalHttp2ConnectionHandler(Http2Connection connection,
+                Http2FrameReader frameReader, Http2FrameWriter frameWriter,
+                Http2FrameListener listener) {
+            super(connection, frameReader, frameWriter, listener);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                Http2Exception.StreamException http2Ex) {
+            super.onStreamError(ctx, cause, http2Ex);
+            Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
+            if (stream == null) {
+                return;
+            }
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            if (streamInfo == null) {
+                return;
+            }
+            streamInfo.childChannel.pipeline().fireExceptionCaught(http2Ex);
+        }
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            // Use a user event in order to circumvent read queue.
```

Comment:
The following was collapsed:

From me:

> I think the use cases you describe make sense...however I just wonder if we can do this without the users having to digest two different logical streams of events. It may be confusing for a client to process the correct amount if the event is received out of order. It would be nice to funnel all of these events through the channelRead mechanism. Would having to be able to be able to add to the front or back of the queue be more confusing? Each message type could be configurable (and have sane defaults).

From @ejona86:

> ChannelRead() just doesn't seem appropriate. Even if we add to the front/back of the queue, we would still delay delivery of the event until they do a read(), which seems a mis-feature. A user event has the exact behavior it seems we need.

Perhaps if I see an example it will become clear. Why would the user expect to get a socket read event unless "they" do a read?  It may make more sense for  "connection" (or "parent" channel) type events (i.e. go_away) where I'm not sure what the typical interaction pattern will look like from a user point of view (the child channels are what the user will be most interested in).  Also using 2 different mechanism to dispatch events and/or dispatching events out of order seems like a confusing default to have.